### ISSUE TITLE
Error: Cannot modify managed objects outside of a write transaction.

### ISSUE THREAD

**Issue by DuranMoodley** [realm/realm-js#2896]
I have a screen that has a toggle switch. Initially , i load my data from my realm database into my state. This gets loaded into my flatlist. I do have an onValue change event for my toggle switch, i want to update my database and refresh my state whenever the user switches the toggle. However, i keep getting "Cannot modify managed objects outside of a write transaction." This happens on both android and ios. I am fairly new to Realm. 

//Retrieving data from realm
 let realmOBJ = new Realm();
 let checklistItems = realmOBJ.objects('CheckpointItems');

setState({
            dataItems:checklistItems//global.ChecklistWithQuestions//dbValuesUnformatted//
   })  

Section that produces the error
function toggleValueChange (index,parentItem,parentIndex)
      {
        let StatusCurrent = parentItem.Questions[index].Status

        let items = [...state.dataItems]

        let newStatus = '';

        if(StatusCurrent === 'Incomplete')
        {
          newStatus = 'Complete'
        }
        else{
          newStatus = 'Incomplete'
        }

        let item = {
          ...parentItem.Questions[index],
          Status:newStatus
        }
       
        parentItem.Questions[index] = item
        items[parentIndex] = parentItem;
        setState({...state,dataItems: items});

        setGlobal({...global,ChecklistWithQuestions:items})

        //Update Realm
        Database.Update(parentItem.ChkpointItemId,parentItem.Questions)
     }

**Comment by kneth** [comment]
Which version of Realm JS do you use? Are you on node.js or RN?

**Comment by steffenagger** [comment]
@DuranMoodley : The error `"Cannot modify managed objects outside of a write transaction."` is thrown when try to update your data outside og a write transaction. It's a bit hard to read your code sample, but generally you need to update your "CheackpointItems" like this:
```js
realm.write(() => {
  checkpointItem.checked = !checkpointItem.checked;
});
```

Hope it helps, otherwise please see: https://realm.io/docs/javascript/latest/#updating-objects

**Comment by mehraj43** [comment]
realm.create('user_details', {
              // user_id: ID,
              _id: await UUIDGenerator.getRandomUUID().then(uuid =>
                uuid.toString(),
              ),
              user_name: userDtls.user_name,
              user_contact: userDtls.user_contact,
              user_address: userDtls.user_address,
            });

im getting Error: Cannot modify managed objects outside of a write transaction.
how can i then get value and assign to the id

**Comment by tomduncalf** [comment]
@mehraj43 Please see the answer above your post, it looks like you are not wrapping your `realm.create` call inside a `realm.write` block. As shown in [the docs](https://www.mongodb.com/docs/realm/sdk/node/examples/read-and-write-data/#create-a-new-object), you need to first open a write transaction before modifying the data.

Something like:

```
realm.write(() => {
  realm.create('user_details', {
    // ...
  });
});
```

Please could you open a new issue for future problems rather than reopening a closed issue, unless it's the exact same issue?

**Comment by mehraj43** [comment]
i have wrapped it just didnt share that part, look

realm.write(async () => {
            realm.create('user_details', {
              _id: await UUIDGenerator.getRandomUUID().then(uuid =>
                uuid.toString(),
              ),
              user_name: userDtls.user_name,
              user_contact: userDtls.user_contact,
              user_address: userDtls.user_address,
            });

**Comment by tomduncalf** [comment]
I don't believe we support async functions in `realm.write`. You'll need to call `await UUIDGenerator.getRandomUUID().then(uuid => uuid.toString())` and store the result in a variable before starting your write transaction (if the UUID generation does indeed need to be async)

**Comment by mehraj43** [comment]
> I don't believe we support async functions in `realm.write`. You'll need to call `await UUIDGenerator.getRandomUUID().then(uuid => uuid.toString())` and store the result in a variable before starting your write transaction (if the UUID generation does indeed need to be async)

Thanks you were really a great help
